---
title: "Nonrandom Patterns of Introgressed Regions"
author: "Matt Glasenapp"
date: "2023-10-25"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(sjPlot)
library(fitdistrplus)
library(ggplot2)
library(ggpubr)
library(plyr)
library(dplyr)
library(fitdistrplus)
library(knitr)
library(tidyr)
library(ggh4x)
library(scales)
```

# R Markdown

## Divergence (dXY) of Introgressed Regions vs. Genome-Wide Background
```{r, echo=FALSE, message=FALSE, warning=FALSE}
# Load the csv files containing the distribution of mean dXY values and read as a one-dimensional vector
# Files for posterior probability > 90

introgression_tract_dxy_dist_file_90 = "/Users/matt/Documents/GitHub/dissertation_chapter_2/data/pixy_dxy/introgression_tracts_mean_dxy_dist_90.csv"

species_tree_tract_dxy_dist_file_90 = "/Users/matt/Documents/GitHub/dissertation_chapter_2/data/pixy_dxy/species_tree_tracts_mean_dxy_dist_90.csv"

dist_introgression_tract_dxy_90 <- scan(introgression_tract_dxy_dist_file_90, what = numeric(), sep = ",")
dist_species_tree_tract_dxy_90 <- scan(species_tree_tract_dxy_dist_file_90, what = numeric(), sep = ",")
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
# Files for posterior probability > 80

introgression_tract_dxy_dist_file_80 = "/Users/matt/Documents/GitHub/dissertation_chapter_2/data/pixy_dxy/introgression_tracts_mean_dxy_dist_80.csv"

species_tree_tract_dxy_dist_file_80 = "/Users/matt/Documents/GitHub/dissertation_chapter_2/data/pixy_dxy/species_tree_tracts_mean_dxy_dist_80.csv"

dist_introgression_tract_dxy_80 <- scan(introgression_tract_dxy_dist_file_80, what = numeric(), sep = ",")
dist_species_tree_tract_dxy_80 <- scan(species_tree_tract_dxy_dist_file_80, what = numeric(), sep = ",")
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
# Add one-dimensional vectors of mean dXY values to a dataframe called df
df_dxy <- data.frame(
  dist_type = factor(c(rep("introgression_tract", 1000), rep("species_tree_tract", 1000), rep("introgression_tract", 1000), rep("species_tree_tract", 1000))),
  mean_divergence = c(dist_introgression_tract_dxy_90, dist_species_tree_tract_dxy_90, dist_introgression_tract_dxy_80, dist_species_tree_tract_dxy_80),
  posterior_probability = factor(c(rep("probability_90", 2000), rep("probability_80", 2000)))
)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
# Reorder levels of posterior_probability
df_dxy$posterior_probability <- factor(df_dxy$posterior_probability, levels = c("probability_90", "probability_80"))
```

``` {r, echo=FALSE, message=FALSE, warning=FALSE}
# Plot histograms of mean absolute nucleotide divergence (dXY)

fig1 <- ggplot(data = df_dxy, aes(x = mean_divergence, fill = dist_type)) +
  geom_histogram(aes(y=after_stat(density)), alpha = 0.4, color = "black", binwidth = 0.0001) +
  geom_density(alpha=0) + 
  facet_wrap(
    ~ posterior_probability,
    ncol = 1,
    scales = "free_x",
    labeller = labeller(
      posterior_probability = c(
        "probability_90" = "Posterior Probability Threshold: 90%",
        "probability_80" = "Posterior Probability Threshold: 80%"
      )
    )
  ) +
  scale_fill_manual(values = c("introgression_tract" = "#A6CEE3", "species_tree_tract" = "#FDBF6F")) +
  labs(
    x = "Divergence",
    y = "Frequency",
    fill = "Dist Type"
  ) + theme_blank() +
  theme(
    legend.position = "none",
    panel.grid = element_blank()
  )
```

```{r, echo=FALSE, message=FALSE, fig.cap = "Figure 1. Distribution of mean absolute nucleotide divergence for introgression tracts and species tree tracts by PhyloNet-HMM posterior probability threshold (90%, 80%)"}
plot(fig1)
```

### Shaprio-Wilk tests and QQ Plots
### Shapiro Test H0: sample is normally distributed 
#### Introgression Tracts: 90% posterior probability threshold
``` {r, echo=FALSE, message=FALSE, fig.cap = "Figure 2. qqplot for mean dXY distribution for introgression tracts at the 90% threshold"}

introgression_tract_dxy_90_shapiro <- shapiro.test(df_dxy$mean_divergence[df_dxy$dist_type == "introgression_tract" & df_dxy$posterior_probability == "probability_90"])

pvalue <- introgression_tract_dxy_90_shapiro$p.value

cat("Shapiro-Wilk test p-value = ", round(pvalue, 3), "\n")

introgression_tract_dxy_90_qqplot <- ggqqplot(df_dxy$mean_divergence[df_dxy$dist_type == "introgression_tract" & df_dxy$posterior_probability == "probability_90"])

plot(introgression_tract_dxy_90_qqplot)
```

#### Species Tree Tracts: 90% posterior probability threshold
``` {r, echo=FALSE, message=FALSE, fig.cap = "Figure 3. qqplot for mean dXY distribution for species tree tracts at the 90% threshold"}

species_tree_tract_dxy_90_shapiro <- shapiro.test(df_dxy$mean_divergence[df_dxy$dist_type == "species_tree_tract" & df_dxy$posterior_probability == "probability_90"])

pvalue <- species_tree_tract_dxy_90_shapiro$p.value

cat("Shapiro-Wilk test p-value = ", round(pvalue, 3), "\n")

species_tree_tract_dxy_90_qqplot <- ggqqplot(df_dxy$mean_divergence[df_dxy$dist_type == "species_tree_tract" & df_dxy$posterior_probability == "probability_90"])

plot(species_tree_tract_dxy_90_qqplot)
```

#### Introgression Tracts: 80% posterior probability threshold
``` {r, echo=FALSE, message=FALSE, fig.cap = "Figure 4. qqplot for mean dXY distribution for introgression tracts at the 80% threshold"}

introgression_tract_dxy_80_shapiro <- shapiro.test(df_dxy$mean_divergence[df_dxy$dist_type == "introgression_tract" & df_dxy$posterior_probability == "probability_80"])

pvalue <- introgression_tract_dxy_80_shapiro$p.value

cat("Shapiro-Wilk test p-value = ", round(pvalue, 3), "\n")

introgression_tract_dxy_80_qqplot <- ggqqplot(df_dxy$mean_divergence[df_dxy$dist_type == "introgression_tract" & df_dxy$posterior_probability == "probability_80"])

plot(introgression_tract_dxy_80_qqplot)
```

#### Species Tree Tracts: 80% posterior probability threshold
``` {r, echo=FALSE, message=FALSE, fig.cap = "Figure 5. qqplot for mean dXY distribution for species tree tracts at the 80% threshold"}

species_tree_tract_dxy_80_shapiro <- shapiro.test(df_dxy$mean_divergence[df_dxy$dist_type == "species_tree_tract" & df_dxy$posterior_probability == "probability_80"])

pvalue <- species_tree_tract_dxy_80_shapiro$p.value

cat("Shapiro-Wilk test p-value = ", round(pvalue, 3), "\n")

species_tree_tract_dxy_80_qqplot <- ggqqplot(df_dxy$mean_divergence[df_dxy$dist_type == "species_tree_tract" & df_dxy$posterior_probability == "probability_80"])

plot(species_tree_tract_dxy_80_qqplot)
```

### Homogeneity of variances 
``` {r, echo=FALSE}
result_dxy <- df_dxy %>%
  group_by(dist_type, posterior_probability) %>%
  summarise(sd_mean_divergence = sd(mean_divergence),
            variance_mean_divergence = sd_mean_divergence^2,
            .groups = "drop") %>%
  arrange(posterior_probability)

# Rename the variables for the table
result_dxy <- result_dxy %>%
  rename(
    "Standard Deviation" = sd_mean_divergence,
    "Variance" = variance_mean_divergence,
    "Posterior Probability" = posterior_probability,
    "Tract Type" = dist_type
  )

# Create vectors to rename values within the table
dist_type_rename_dxy <- c("introgression_tract" = "Introgression", "species_tree_tract" = "Species Tree")
posterior_probability_rename_dxy <- c("probability_90" = "90%", "probability_80" = "80%")

# Apply the renaming to the data frame
result <- result_dxy %>%
  mutate(
    `Tract Type` = case_when(
      `Tract Type` %in% names(dist_type_rename_dxy) ~ dist_type_rename_dxy[`Tract Type`],
      TRUE ~ as.character(`Tract Type`)
    ),
    `Posterior Probability` = case_when(
      `Posterior Probability` %in% names(posterior_probability_rename_dxy) ~ posterior_probability_rename_dxy[`Posterior Probability`],
      TRUE ~ as.character(`Posterior Probability`)
    )
  )

# Print the kable table with customized values
knitr::kable(result_dxy, digits = 8, align = "c", format.args = list(scientific = FALSE))
```


### F-test to assess homogeneity of variances for dXY between introgression tracts and species tree tracts
### F Test H0: Ratio of variances is equal
``` {r, echo=FALSE, message=FALSE, warning=FALSE}
# Perform variance tests
var_test_90_dxy <- var.test(df_dxy$mean_divergence[df_dxy$dist_type == "introgression_tract" & df_dxy$posterior_probability == "probability_90"],
                        df_dxy$mean_divergence[df_dxy$dist_type == "species_tree_tract" & df_dxy$posterior_probability == "probability_90"])

var_test_80_dxy <- var.test(df_dxy$mean_divergence[df_dxy$dist_type == "introgression_tract" & df_dxy$posterior_probability == "probability_80"],
                        df_dxy$mean_divergence[df_dxy$dist_type == "species_tree_tract" & df_dxy$posterior_probability == "probability_80"])

log_var_test_90_dxy <- var.test(log(df_dxy$mean_divergence[df_dxy$dist_type == "introgression_tract" & df_dxy$posterior_probability == "probability_90"]),
                            log(df_dxy$mean_divergence[df_dxy$dist_type == "species_tree_tract" & df_dxy$posterior_probability == "probability_90"]))

log_var_test_80_dxy <- var.test(log(df_dxy$mean_divergence[df_dxy$dist_type == "introgression_tract" & df_dxy$posterior_probability == "probability_80"]),
                            log(df_dxy$mean_divergence[df_dxy$dist_type == "species_tree_tract" & df_dxy$posterior_probability == "probability_80"]))

# Function to format the confidence interval with three digits
format_confidence_interval <- function(conf.int) {
  return(sprintf("[%.3f, %.3f]", conf.int[1], conf.int[2]))
}

# Calculate the ratio of variances
ratio_of_variances_dxy <- c(var_test_90_dxy$estimate, var_test_80_dxy$estimate, log_var_test_90_dxy$estimate, log_var_test_80_dxy$estimate)

# Create a data frame with test results, including the variance ratio
results_dxy <- data.frame(
  "Posterior Probability" = c("90%", "80%", "log(90%)", "log(80%)"),
    "Variance Ratio" = ratio_of_variances_dxy,
  "p value" = c(var_test_90_dxy$p.value, var_test_80_dxy$p.value, log_var_test_90_dxy$p.value, log_var_test_80_dxy$p.value),
  "Confidence Interval" = sapply(list(var_test_90_dxy$conf.int, var_test_80_dxy$conf.int, log_var_test_90_dxy$conf.int, log_var_test_80_dxy$conf.int), format_confidence_interval)
)

# Create a knitr table
kable(results_dxy, format = "markdown", digits = 3, align ="c", col.names = c("Posterior Probability", "Variance Ratio", "p", "Confidence Interval"))

```

### Welch Two Sample t-test
``` {r, echo=FALSE, message=FALSE, warning=FALSE}
subset_90_dxy = df_dxy[df_dxy$posterior_probability == "probability_90", ]
results_90_dxy <- t.test(mean_divergence~dist_type, data=subset_90_dxy, paired=FALSE, var.eq=F, alternative="two.sided")

subset_80_dxy = df_dxy[df_dxy$posterior_probability == "probability_80", ]
results_80_dxy <- t.test(mean_divergence~dist_type, data=subset_80_dxy, paired=FALSE, var.eq=T, alternative="two.sided")

# Create a data frame to store the t-test results
t_test_results_dxy <- data.frame(
  posterior_probability = c("90%", "80%"),
  t = c(results_90_dxy$statistic, results_80_dxy$statistic),
  df = c(results_90_dxy$parameter, results_80_dxy$parameter),
  p_value = c(results_90_dxy$p.value, results_80_dxy$p.value),
  conf_interval = c(paste0("[", round(results_90_dxy$conf.int[1], 4), ", ", round(results_90_dxy$conf.int[2], 4), "]"),
                   paste0("[", round(results_80_dxy$conf.int[1], 4), ", ", round(results_80_dxy$conf.int[2], 4), "]")
  )
)

# Create a kable table
kable(t_test_results_dxy, format = "markdown", align = "c", digits = 3, col.names = c("Posterior Probability", "t", "df", "p", "Confidence Interval"))
```

### Figure for manuscript
```{r, echo=FALSE, message=FALSE}
fig6 <- ggplot(data = df_dxy, aes(x = mean_divergence, fill = dist_type)) +
  geom_density(aes(y=after_stat(density))) + 
  facet_wrap(
    ~ posterior_probability,
    ncol = 1,
    scales = "free_y",
    #scales = "free_x",
    labeller = labeller(
      posterior_probability = c(
        "probability_90" = "Posterior Probability Threshold: 90%",
        "probability_80" = "Posterior Probability Threshold: 80%"
      )
    )
  ) +
  scale_fill_manual(values = c("introgression_tract" = "#A6CEE3", "species_tree_tract" = "#999999"), labels = c("introgression_tract" = "Introgression Tracts", "species_tree_tract" = "Genome-Wide Background")
) + labs(
  x = expression(
      atop(
        paste("Mean ", italic("H. pulcherrimus"), " - ", italic("S. fragilis")),
        paste("Absolute Nucleotide Divergence (", italic("d")[XY], ")")
      )
    ),
    y = "Probability Density",
    fill = "Dist Type"
  ) + theme_blank() +
  theme(
    axis.title = element_text(size = 16),
    legend.position = "right",
    panel.grid = element_blank(), 
    strip.text = element_text(size = 14),
    legend.text = element_text(size = 12)
  ) + 
  guides(fill = guide_legend(title = NULL))
```

```{r, echo=FALSE, message=FALSE, fig.cap = "Figure 6. Distribution of mean absolute nucleotide divergence for introgression tracts and species tree tracts by PhyloNet-HMM posterior probability threshold (90%, 80%)"}
plot(fig6)
```


```{r, echo=FALSE, message=FALSE}
#setwd("/Users/matt/Documents/GitHub/dissertation_chapter_2/data/pixy_dxy/")
#ggsave(filename = "dXY.eps", plot = fig2)
#ggsave(filename = "dXY.png", plot = fig2)
#ggsave(filename = "dXY.svg", plot = fig2)
```

``` {r, echo=FALSE}
#model1 <- lm(dist_species_tree_tract_dxy_80 ~ 1)
#model2 <- lm(dist_species_tree_tract_dxy_90 ~ 1)

# Calculate the confidence interval
#confint(model1, level=0.95)
#confint(model2, level=0.95)

#mean(dist_introgression_tract_dxy_80)
#mean(dist_introgression_tract_dxy_90)
```

## Gene Density of Introgressed Regions vs. Genome-Wide Background
``` {r, echo=FALSE}
# Load the csv files containing the distribution of gene counts and read as a one-dimensional vector

# Define the csv files with list of gene counts at 90% posterior probability threshold
introgression_tract_gene_count_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/introgression_tract_gene_count_90.csv"
species_tree_tract_gene_count_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/species_tree_tract_gene_count_90.csv"

# Define the csv files with list of gene counts at 80% posterior probability threshold
introgression_tract_gene_count_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/introgression_tract_gene_count_80.csv"
species_tree_tract_gene_count_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/species_tree_tract_gene_count_80.csv"

# Read the CSV file as a one-dimensional vector
dist_introgression_tract_genes_90 <- scan(introgression_tract_gene_count_90, what = numeric(), sep = ",")
dist_species_tree_tract_genes_90 <- scan(species_tree_tract_gene_count_90, what = numeric(), sep = ",")
dist_introgression_tract_genes_80 <- scan(introgression_tract_gene_count_80, what = numeric(), sep = ",")
dist_species_tree_tract_genes_80 <- scan(species_tree_tract_gene_count_80, what = numeric(), sep = ",")

mb_90 = 3.747392
mb_80 = 21.820425

dist_introgression_tract_genes_90 <- dist_introgression_tract_genes_90 / mb_90
dist_species_tree_tract_genes_90 <- dist_species_tree_tract_genes_90 / mb_90
dist_introgression_tract_genes_80 <- dist_introgression_tract_genes_80 / mb_80
dist_species_tree_tract_genes_80 <- dist_species_tree_tract_genes_80 / mb_80
```

``` {r, echo=FALSE}
# Add one-dimensional vectors of mean dXY values to a dataframe called df
df_gene_density <- data.frame(
  dist_type = factor(c(rep("introgression_tract", 1000), rep("species_tree_tract", 1000), rep("introgression_tract", 1000), rep("species_tree_tract", 1000))),
  gene_count = c(dist_introgression_tract_genes_90, dist_species_tree_tract_genes_90, dist_introgression_tract_genes_80, dist_species_tree_tract_genes_80),
  posterior_probability = factor(c(rep("probability_90", 2000), rep("probability_80", 2000)))
)

# Reorder levels of posterior_probability
df_gene_density$posterior_probability <- factor(df_gene_density$posterior_probability, levels = c("probability_90", "probability_80"))
```

``` {r, echo = FALSE}
# Plot histograms of protein-coding genes/mb 
fig7 <- ggplot(data = df_gene_density, aes(x = gene_count, fill = dist_type)) +
  geom_histogram(aes(y=after_stat(density)), alpha = 0.4, color = "black", binwidth = 1) +
  geom_density(alpha=0) + 
  facet_wrap(
    ~ posterior_probability,
    ncol = 1,
    scales = "free_x",
    labeller = labeller(
      posterior_probability = c(
        "probability_90" = "Posterior Probability Threshold: 90%",
        "probability_80" = "Posterior Probability Threshold: 80%"
      )
    )
  ) +
  scale_fill_manual(values = c("introgression_tract" = "#A6CEE3", "species_tree_tract" = "#FDBF6F")) +
  labs(
    x = "Divergence",
    y = "Frequency",
    fill = "Dist Type"
  ) + theme_blank() +
  theme(
    legend.position = "none",
    panel.grid = element_blank()
  )
```

```{r, echo=FALSE, message=FALSE, fig.cap = "Figure 7. Distribution of the number of overlapping protein-coding genes per Mb for introgression tracts and species tree tracts by PhyloNet-HMM posterior probability threshold (90%, 80%)"}
plot(fig7)
```

### Introgression Tract Gene Count: 90% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 8. Normal fit of gene counts for introgression tracts at the 90% posterior probability threshold"}
fit.norm_introgressed_genes_90 = fitdist(df_gene_density$gene_count[df_gene_density$dist_type == "introgression_tract" & df_gene_density$posterior_probability == "probability_90"], "norm")
plot(fit.norm_introgressed_genes_90)
```

### Species Tree Tract Gene Count: 90% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 9. Normal fit of gene counts for species tree tracts at the 90% posterior probability threshold"}
fit.norm_species_tree_genes_90 = fitdist(df_gene_density$gene_count[df_gene_density$dist_type == "species_tree_tract" & df_gene_density$posterior_probability == "probability_90"], "norm")
plot(fit.norm_species_tree_genes_90)
```

### Introgression Tract Gene Count: 80% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 10. Normal fit of gene counts for introgression tracts at the 80% posterior probability threshold"}
fit.norm_introgressed_genes_80 = fitdist(df_gene_density$gene_count[df_gene_density$dist_type == "introgression_tract" & df_gene_density$posterior_probability == "probability_80"], "norm")
plot(fit.norm_introgressed_genes_80)
```

### Species Tree Tract Gene Count: 80% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 11. Normal fit of gene counts for species tree tracts at the 80% posterior probability threshold"}
fit.norm_species_tree_genes_80 = fitdist(df_gene_density$gene_count[df_gene_density$dist_type == "species_tree_tract" & df_gene_density$posterior_probability == "probability_80"], "norm")
plot(fit.norm_species_tree_genes_80)
```


### Normal vs. log normal fit for Gene Counts
```{r, echo=FALSE}
fit.lnorm_introgressed_genes_90 = fitdist(df_gene_density$gene_count[df_gene_density$dist_type == "introgression_tract" & df_gene_density$posterior_probability == "probability_90"], "lnorm")

fit.lnorm_species_tree_genes_90 = fitdist(df_gene_density$gene_count[df_gene_density$dist_type == "species_tree_tract" & df_gene_density$posterior_probability == "probability_90"], "lnorm")

fit.lnorm_introgressed_genes_80 = fitdist(df_gene_density$gene_count[df_gene_density$dist_type == "introgression_tract" & df_gene_density$posterior_probability == "probability_80"], "lnorm")

fit.lnorm_species_tree_genes_80 = fitdist(df_gene_density$gene_count[df_gene_density$dist_type == "species_tree_tract" & df_gene_density$posterior_probability == "probability_80"], "lnorm")

# Create a data frame to store the t-test results
gene_count_aic <- data.frame(
  tract_type = c("Introgression", "Introgression", "Introgression", "Introgression", "Species Tree","Species Tree","Species Tree","Species Tree"),
  posterior_probability = c("90%", "90%", "80%", "80%", "90%", "90%", "80%", "80%"),
  fit = c("Normal", "Log Normal", "Normal", "Log Normal", "Normal", "Log Normal", "Normal", "Log Normal"),
  aic = c(fit.norm_introgressed_genes_90$aic, fit.lnorm_introgressed_genes_90$aic, fit.norm_introgressed_genes_80$aic, fit.lnorm_introgressed_genes_80$aic, fit.norm_species_tree_genes_90$aic, fit.lnorm_species_tree_genes_90$aic, fit.norm_species_tree_genes_90$aic, fit.lnorm_species_tree_genes_90$aic))

# Create a kable table
kable(gene_count_aic, format = "markdown", align = "c", digits = 1, col.names = c("Tract Type", "Posterior Probability", "Fit", "aic"))
```

### Homogeneity of variances for Gene Count
``` {r, echo=FALSE}
result_gene_density <- df_gene_density %>%
  group_by(dist_type, posterior_probability) %>%
  summarise(sd_gene_count = sd(gene_count),
            variance_gene_count = sd_gene_count^2,
            .groups = "drop") %>%
  arrange(posterior_probability)

# Rename the variables for the table
result_gene_density <- result_gene_density %>%
  rename(
    "Standard Deviation" = sd_gene_count,
    "Variance" = variance_gene_count,
    "Posterior Probability" = posterior_probability,
    "Tract Type" = dist_type
  )

# Create vectors to rename values within the table
dist_type_rename_gene_count <- c("introgression_tract" = "Introgression", "species_tree_tract" = "Species Tree")
posterior_probability_rename_gene_count <- c("probability_90" = "90%", "probability_80" = "80%")

# Apply the renaming to the data frame
result <- result_gene_density %>%
  mutate(
    `Tract Type` = case_when(
      `Tract Type` %in% names(dist_type_rename_gene_count) ~ dist_type_rename_gene_count[`Tract Type`],
      TRUE ~ as.character(`Tract Type`)
    ),
    `Posterior Probability` = case_when(
      `Posterior Probability` %in% names(posterior_probability_rename_gene_count) ~ posterior_probability_rename_gene_count[`Posterior Probability`],
      TRUE ~ as.character(`Posterior Probability`)
    )
  )

# Print the kable table with customized values
knitr::kable(result, digits = 8, align = "c", format.args = list(scientific = FALSE))
```

### F-test to assess homogeneity of variances between gene counts for introgression tracts and species tree tracts
#### F Test H0: Ratio of variances is equal
``` {r, echo=FALSE, message=FALSE, warning=FALSE}
# Perform variance tests
var_test_90_gene_density <- var.test(df_gene_density$gene_count[df_gene_density$dist_type == "introgression_tract" &
                                                                df_gene_density$posterior_probability == "probability_90"],
                                   df_gene_density$gene_count[df_gene_density$dist_type == "species_tree_tract" &
                                                                df_gene_density$posterior_probability == "probability_90"])

var_test_80_gene_density <- var.test(df_gene_density$gene_count[df_gene_density$dist_type == "introgression_tract" &
                                                                df_gene_density$posterior_probability == "probability_80"],
                                   df_gene_density$gene_count[df_gene_density$dist_type == "species_tree_tract" &
                                                                df_gene_density$posterior_probability == "probability_80"])

log_var_test_90_gene_density <- var.test(log(df_gene_density$gene_count[df_gene_density$dist_type == "introgression_tract" &
                                                                          df_gene_density$posterior_probability == "probability_90"]),
                                         log(df_gene_density$gene_count[df_gene_density$dist_type == "species_tree_tract" &
                                                                          df_gene_density$posterior_probability == "probability_90"]))

log_var_test_80_gene_density <- var.test(log(df_gene_density$gene_count[df_gene_density$dist_type == "introgression_tract" &
                                                                          df_gene_density$posterior_probability == "probability_80"]),

                                         log(df_gene_density$gene_count[df_gene_density$dist_type == "species_tree_tract" &
                                                                               df_gene_density$posterior_probability ==
                                                                          "probability_80"]))

# Function to format the confidence interval with three digits
format_confidence_interval <- function(conf.int) {
  return(sprintf("[%.3f, %.3f]", conf.int[1], conf.int[2]))
}

# Calculate the ratio of variances
ratio_of_variances_gene_density <- c(var_test_90_gene_density$estimate, var_test_80_gene_density$estimate, log_var_test_90_gene_density$estimate, log_var_test_80_gene_density$estimate)

# Create a data frame with test results, including the variance ratio
results_gene_density <- data.frame(
  "Posterior Probability" = c("90%", "80%", "log(90%)", "log(80%)"),
    "Variance Ratio" = ratio_of_variances_gene_density,
  "p value" = c(var_test_90_gene_density$p.value, var_test_80_gene_density$p.value, log_var_test_90_gene_density$p.value, log_var_test_80_gene_density$p.value),
  "Confidence Interval" = sapply(list(var_test_90_gene_density$conf.int, var_test_80_gene_density$conf.int, log_var_test_90_gene_density$conf.int, log_var_test_80_gene_density$conf.int), format_confidence_interval)
)

# Create a knitr table
kable(results_gene_density, format = "markdown", digits = 3, align ="c", col.names = c("Posterior Probability", "Variance Ratio", "p", "Confidence Interval"))

```

### Welch Two Sample t-test
``` {r, echo=FALSE, message=FALSE, warning=FALSE}
subset_90_gene_density = df_gene_density[df_gene_density$posterior_probability == "probability_90", ]
results_90_gene_density <- t.test(gene_count~dist_type, data=subset_90_gene_density, paired=FALSE, var.eq=F, alternative="two.sided")

subset_80_gene_density = df_gene_density[df_gene_density$posterior_probability == "probability_80", ]
results_80_gene_density <- t.test(gene_count~dist_type, data=subset_80_gene_density, paired=FALSE, var.eq=T, alternative="two.sided")

# Create a data frame to store the t-test results
t_test_results_gene_density <- data.frame(
  posterior_probability = c("90%", "80%"),
  t = c(results_90_gene_density$statistic, results_80_gene_density$statistic),
  df = c(results_90_gene_density$parameter, results_80_gene_density$parameter),
  p_value = c(results_90_gene_density$p.value, results_80_gene_density$p.value),
  conf_interval = c(paste0("[", round(results_90_gene_density$conf.int[1], 4), ", ", round(results_90_gene_density$conf.int[2], 4), "]"),
                   paste0("[", round(results_80_gene_density$conf.int[1], 4), ", ", round(results_80_gene_density$conf.int[2], 4), "]")
  )
)

# Create a kable table
kable(t_test_results_gene_density, format = "markdown", align = "c", digits = 3, col.names = c("Posterior Probability", "t", "df", "p", "Confidence Interval"))
```

``` {r, echo=FALSE}
fig12 <- ggplot(data = df_gene_density, aes(x = gene_count, fill = dist_type)) +
  geom_density(aes(y=after_stat(density)),alpha=0.8) + 
  facet_wrap(
    ~ posterior_probability,
    ncol = 1,
    scales = "free",
    #scales = "free_y",
    #scales = "free_x",
    labeller = labeller(
      posterior_probability = c(
        "probability_90" = "Posterior Probability Threshold: 90%",
        "probability_80" = "Posterior Probability Threshold: 80%"
      )
    )
  ) +
  scale_fill_manual(values = c("introgression_tract" = "#A6CEE3", "species_tree_tract" = "#999999"), labels = c("introgression_tract" = "Introgression Tracts", "species_tree_tract" = "Genome-Wide Background")
  ) + labs(
    x = "Protein-Coding Genes / Mb",
    y = "Probability Density",
    fill = "Dist Type"
  ) + theme_bw() +
  theme(
    axis.title = element_text(size = 16),
    legend.text = element_text(size = 12),
    strip.text.x = element_text(size = 12),
    panel.grid = element_blank(), 
    legend.position = "top",
    axis.text = element_text(size = 12)
  ) + 
  guides(fill = guide_legend(title = NULL))
```

### Figure for Manuscript
```{r, echo=FALSE, fig.cap = "Figure 10. Distribution of gene counts for introgression tracts and species tree tracts by PhyloNet-HMM posterior probability threshold (90%, 80%)"}
plot(fig12)
```

## Percent Coding for Introgressed Regions vs. Genome-Wide Background

``` {r, echo = FALSE}
# Define the csv files with list of base counts at 90% posterior probability threshold
introgression_tract_base_count_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/introgression_tract_base_count_90.csv"
species_tree_tract_base_count_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/species_tree_tract_base_count_90.csv"

# Define the csv files with list of base counts at 80% posterior probability threshold
introgression_tract_base_count_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/introgression_tract_base_count_80.csv"
species_tree_tract_base_count_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/species_tree_tract_base_count_80.csv"

# Read the CSV file as a one-dimensional vector
dist_introgression_tract_bases_90 <- scan(introgression_tract_base_count_90, what = numeric(), sep = ",")
dist_species_tree_tract_bases_90 <- scan(species_tree_tract_base_count_90, what = numeric(), sep = ",")
dist_introgression_tract_bases_80 <- scan(introgression_tract_base_count_80, what = numeric(), sep = ",")
dist_species_tree_tract_bases_80 <- scan(species_tree_tract_base_count_80, what = numeric(), sep = ",")

total_bases_90 = 3747392
total_bases_80 = 21820425

dist_introgression_tract_bases_90 <- dist_introgression_tract_bases_90 / total_bases_90 * 100
dist_introgression_tract_bases_80 <- dist_introgression_tract_bases_80 / total_bases_80 * 100
dist_species_tree_tract_bases_90 <- dist_species_tree_tract_bases_90 / total_bases_90 * 100
dist_species_tree_tract_bases_80 <- dist_species_tree_tract_bases_80 / total_bases_80 * 100
```

``` {r, echo = FALSE}
# Add one-dimensional vectors of mean dXY values to a dataframe called df
df_coding <- data.frame(
  dist_type = factor(c(rep("introgression_tract", 1000), rep("species_tree_tract", 1000), rep("introgression_tract", 1000), rep("species_tree_tract", 1000))),
  base_count = c(dist_introgression_tract_bases_90, dist_species_tree_tract_bases_90, dist_introgression_tract_bases_80, dist_species_tree_tract_bases_80),
  posterior_probability = factor(c(rep("probability_90", 2000), rep("probability_80", 2000)))
)

# Reorder levels of posterior_probability
df_coding$posterior_probability <- factor(df_coding$posterior_probability, levels = c("probability_90", "probability_80"))
```

### Introgression Tract Percent Coding: 90% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 8. Normal fit of gene counts for introgression tracts at the 90% posterior probability threshold"}
fit.norm_introgressed_genes_90 = fitdist(df_coding$base_count[df_coding$dist_type == "introgression_tract" & df_coding$posterior_probability == "probability_90"], "norm")
plot(fit.norm_introgressed_genes_90)
```

### Species Tree Tract Percent Coding: 90% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 9. Normal fit of gene counts for species tree tracts at the 90% posterior probability threshold"}
fit.norm_species_tree_genes_90 = fitdist(df_coding$base_count[df_coding$dist_type == "species_tree_tract" & df_coding$posterior_probability == "probability_90"], "norm")
plot(fit.norm_species_tree_genes_90)
```

### Introgression Tract Percent Coding: 80% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 10. Normal fit of gene counts for introgression tracts at the 80% posterior probability threshold"}
fit.norm_introgressed_genes_80 = fitdist(df_coding$base_count[df_coding$dist_type == "introgression_tract" & df_coding$posterior_probability == "probability_80"], "norm")
plot(fit.norm_introgressed_genes_80)
```

### Species Tree Tract Percent Coding: 80% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 11. Normal fit of gene counts for species tree tracts at the 80% posterior probability threshold"}
fit.norm_species_tree_genes_80 = fitdist(df_coding$base_count[df_coding$dist_type == "species_tree_tract" & df_coding$posterior_probability == "probability_80"], "norm")
plot(fit.norm_species_tree_genes_80)
```

### Homogeneity of variances for Percent Coding
``` {r, echo=FALSE}
result_coding <- df_coding %>%
  group_by(dist_type, posterior_probability) %>%
  summarise(sd_base_count = sd(base_count),
            variance_base_count = sd_base_count^2,
            .groups = "drop") %>%
  arrange(posterior_probability)

# Rename the variables for the table
result_coding <- result_coding %>%
  rename(
    "Standard Deviation" = sd_base_count,
    "Variance" = variance_base_count,
    "Posterior Probability" = posterior_probability,
    "Tract Type" = dist_type
  )

# Create vectors to rename values within the table
dist_type_rename_base_count <- c("introgression_tract" = "Introgression", "species_tree_tract" = "Species Tree")
posterior_probability_rename_base_count <- c("probability_90" = "90%", "probability_80" = "80%")

# Apply the renaming to the data frame
result <- result_coding %>%
  mutate(
    `Tract Type` = case_when(
      `Tract Type` %in% names(dist_type_rename_base_count) ~ dist_type_rename_base_count[`Tract Type`],
      TRUE ~ as.character(`Tract Type`)
    ),
    `Posterior Probability` = case_when(
      `Posterior Probability` %in% names(posterior_probability_rename_base_count) ~ posterior_probability_rename_base_count[`Posterior Probability`],
      TRUE ~ as.character(`Posterior Probability`)
    )
  )

# Print the kable table with customized values
knitr::kable(result, digits = 8, align = "c", format.args = list(scientific = FALSE))
```

### F-test to assess homogeneity of variances between percent coding for introgression tracts and species tree tracts
#### F Test H0: Ratio of variances is equal
``` {r, echo=FALSE, message=FALSE, warning=FALSE}
# Perform variance tests
var_test_90_coding <- var.test(df_coding$base_count[df_coding$dist_type == "introgression_tract" &
                                                                  df_coding$posterior_probability == "probability_90"],
                                     df_coding$base_count[df_coding$dist_type == "species_tree_tract" &
                                                                  df_coding$posterior_probability == "probability_90"])

var_test_80_coding <- var.test(df_coding$base_count[df_coding$dist_type == "introgression_tract" &
                                                                  df_coding$posterior_probability == "probability_80"],
                                     df_coding$base_count[df_coding$dist_type == "species_tree_tract" &
                                                                  df_coding$posterior_probability == "probability_80"])

log_var_test_90_coding <- var.test(log(df_coding$base_count[df_coding$dist_type == "introgression_tract" &
                                                                          df_coding$posterior_probability == "probability_90"]),
                                         log(df_coding$base_count[df_coding$dist_type == "species_tree_tract" &
                                                                          df_coding$posterior_probability == "probability_90"]))

log_var_test_80_coding <- var.test(log(df_coding$base_count[df_coding$dist_type == "introgression_tract" &
                                                                          df_coding$posterior_probability == "probability_80"]),
                                         
                                         log(df_coding$base_count[df_coding$dist_type == "species_tree_tract" &
                                                                          df_coding$posterior_probability ==
                                                                          "probability_80"]))

# Function to format the confidence interval with three digits
format_confidence_interval <- function(conf.int) {
  return(sprintf("[%.3f, %.3f]", conf.int[1], conf.int[2]))
}

# Calculate the ratio of variances
ratio_of_variances_coding <- c(var_test_90_coding$estimate, var_test_80_coding$estimate, log_var_test_90_coding$estimate, log_var_test_80_coding$estimate)

# Create a data frame with test results, including the variance ratio
results_coding <- data.frame(
  "Posterior Probability" = c("90%", "80%", "log(90%)", "log(80%)"),
  "Variance Ratio" = ratio_of_variances_coding,
  "p value" = c(var_test_90_coding$p.value, var_test_80_coding$p.value, log_var_test_90_coding$p.value, log_var_test_80_coding$p.value),
  "Confidence Interval" = sapply(list(var_test_90_coding$conf.int, var_test_80_coding$conf.int, log_var_test_90_coding$conf.int, log_var_test_80_coding$conf.int), format_confidence_interval)
)

# Create a knitr table
kable(results_coding, format = "markdown", digits = 3, align ="c", col.names = c("Posterior Probability", "Variance Ratio", "p", "Confidence Interval"))

```

### Figure for Manuscript
``` {r, echo = FALSE}
fig <- ggplot(data = df_coding, aes(x = base_count, fill = dist_type)) +
  geom_density(aes(y=after_stat(density)), alpha=0.8) + 
  facet_wrap(
    ~ posterior_probability,
    ncol = 1,
    scales = "free",
    #scales = "free_y",
    #scales = "free_x",
    labeller = labeller(
      posterior_probability = c(
        "probability_90" = "Posterior Probability Threshold: 90%",
        "probability_80" = "Posterior Probability Threshold: 80%"
      )
    )
  ) +
  scale_fill_manual(values = c("introgression_tract" = "#A6CEE3", "species_tree_tract" = "#999999"), labels = c("introgression_tract" = "Introgression Tracts", "species_tree_tract" = "Genome-Wide Background")
  ) + labs(
    x = "Percent Coding",
    y = "Probability Density",
    fill = "Dist Type"
  ) + theme_bw() +
  theme(
    axis.title = element_text(size = 16),
    legend.text = element_text(size = 12),
    strip.text.x = element_text(size = 12),
    panel.grid = element_blank(), 
    legend.position = "top",
    axis.text = element_text(size = 12)
  ) + 
  guides(fill = guide_legend(title = NULL))
```

```{r, echo=FALSE, message=FALSE, fig.cap = "Figure X. Percentage of coding bases for introgression tracts greater than 10 kb in length and the genome-wide background."}
plot(fig)
```

## Number of Positively Selected Genes for Introgressed Regions vs. Genome-Wide Background
``` {r, echo = FALSE}
# Define the csv files with list of gene and base countscounts 
introgression_tract_psg_count_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/introgression_tract_psg_count_90.csv"
species_tree_tract_psg_count_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/species_tree_tract_psg_count_90.csv"
introgression_tract_psg_count_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/introgression_tract_psg_count_80.csv"
species_tree_tract_psg_count_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/gene_density/species_tree_tract_psg_count_80.csv"

# Read the CSV file as a one-dimensional vector
dist_introgression_tract_psg_90 <- scan(introgression_tract_psg_count_90, what = numeric(), sep = ",")
dist_species_tree_tract_psg_90 <- scan(species_tree_tract_psg_count_90, what = numeric(), sep = ",")
dist_introgression_tract_psg_80 <- scan(introgression_tract_psg_count_80, what = numeric(), sep = ",")
dist_species_tree_tract_psg_80 <- scan(species_tree_tract_psg_count_80, what = numeric(), sep = ",")

mb_90 = 3.747392
mb_80 = 21.820425

dist_introgression_tract_psg_90 <- dist_introgression_tract_psg_90 / mb_90
dist_species_tree_tract_psg_90 <- dist_species_tree_tract_psg_90 / mb_90
dist_introgression_tract_psg_80 <- dist_introgression_tract_psg_80 / mb_80
dist_species_tree_tract_psg_80 <- dist_species_tree_tract_psg_80 / mb_80
```

``` {r, echo=FALSE}
df_psg <- data.frame(
  dist_type = factor(c(rep("introgression_tract", 1000), rep("species_tree_tract", 1000), rep("introgression_tract", 1000), rep("species_tree_tract", 1000))),
  psg_count = c(dist_introgression_tract_psg_90, dist_species_tree_tract_psg_90, dist_introgression_tract_psg_80, dist_species_tree_tract_psg_80),
  posterior_probability = factor(c(rep("probability_90", 2000), rep("probability_80", 2000)))
)

# Reorder levels of posterior_probability
df_psg$posterior_probability <- factor(df_psg$posterior_probability, levels = c("probability_90", "probability_80"))
```

### Introgression Tract PSG Count: 90% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 8. Normal fit of gene counts for introgression tracts at the 90% posterior probability threshold"}
fit.norm_introgressed_genes_90 = fitdist(df_psg$psg_count[df_psg$dist_type == "introgression_tract" & df_psg$posterior_probability == "probability_90"], "norm")
plot(fit.norm_introgressed_genes_90)
```

### Species Tree Tract PSG Count: 90% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 9. Normal fit of gene counts for species tree tracts at the 90% posterior probability threshold"}
fit.norm_species_tree_genes_90 = fitdist(df_psg$psg_count[df_psg$dist_type == "species_tree_tract" & df_psg$posterior_probability == "probability_90"], "norm")
plot(fit.norm_species_tree_genes_90)
```

### Introgression Tract PSG Count: 80% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 10. Normal fit of gene counts for introgression tracts at the 80% posterior probability threshold"}
fit.norm_introgressed_genes_80 = fitdist(df_psg$psg_count[df_psg$dist_type == "introgression_tract" & df_psg$posterior_probability == "probability_80"], "norm")
plot(fit.norm_introgressed_genes_80)
```

### Species Tree Tract PSG Count: 80% posterior probability threshold
```{r, echo=FALSE, fig.cap = "Figure 11. Normal fit of gene counts for species tree tracts at the 80% posterior probability threshold"}
fit.norm_species_tree_genes_80 = fitdist(df_psg$psg_count[df_psg$dist_type == "species_tree_tract" & df_psg$posterior_probability == "probability_80"], "norm")
plot(fit.norm_species_tree_genes_80)
```


### Homogeneity of variances for PSG Count
``` {r, echo=FALSE}
result_psg <- df_psg %>%
  group_by(dist_type, posterior_probability) %>%
  summarise(sd_psg_count = sd(psg_count),
            variance_psg_count = sd_psg_count^2,
            .groups = "drop") %>%
  arrange(posterior_probability)

# Rename the variables for the table
result_psg <- result_psg %>%
  rename(
    "Standard Deviation" = sd_psg_count,
    "Variance" = variance_psg_count,
    "Posterior Probability" = posterior_probability,
    "Tract Type" = dist_type
  )

# Create vectors to rename values within the table
dist_type_rename_psg_count <- c("introgression_tract" = "Introgression", "species_tree_tract" = "Species Tree")
posterior_probability_rename_psg_count <- c("probability_90" = "90%", "probability_80" = "80%")

# Apply the renaming to the data frame
result <- result_psg %>%
  mutate(
    `Tract Type` = case_when(
      `Tract Type` %in% names(dist_type_rename_psg_count) ~ dist_type_rename_psg_count[`Tract Type`],
      TRUE ~ as.character(`Tract Type`)
    ),
    `Posterior Probability` = case_when(
      `Posterior Probability` %in% names(posterior_probability_rename_psg_count) ~ posterior_probability_rename_psg_count[`Posterior Probability`],
      TRUE ~ as.character(`Posterior Probability`)
    )
  )

# Print the kable table with customized values
knitr::kable(result, digits = 8, align = "c", format.args = list(scientific = FALSE))
```

### F-test to assess homogeneity of variances between PSG counts for introgression tracts and species tree tracts
#### F Test H0: Ratio of variances is equal
``` {r, echo=FALSE, message=FALSE, warning=FALSE}
# Perform variance tests
var_test_90_psg <- var.test(df_psg$psg_count[df_psg$dist_type == "introgression_tract" &
                                                                  df_psg$posterior_probability == "probability_90"],
                                     df_psg$psg_count[df_psg$dist_type == "species_tree_tract" &
                                                                  df_psg$posterior_probability == "probability_90"])

var_test_80_psg <- var.test(df_psg$psg_count[df_psg$dist_type == "introgression_tract" &
                                                                  df_psg$posterior_probability == "probability_80"],
                                     df_psg$psg_count[df_psg$dist_type == "species_tree_tract" &
                                                                  df_psg$posterior_probability == "probability_80"])

log_var_test_90_psg <- var.test(log(df_psg$psg_count[df_psg$dist_type == "introgression_tract" &
                                                                          df_psg$posterior_probability == "probability_90"]),
                                         log(df_psg$psg_count[df_psg$dist_type == "species_tree_tract" &
                                                                          df_psg$posterior_probability == "probability_90"]))

log_var_test_80_psg <- var.test(log(df_psg$psg_count[df_psg$dist_type == "introgression_tract" &
                                                                          df_psg$posterior_probability == "probability_80"]),
                                         
                                         log(df_psg$psg_count[df_psg$dist_type == "species_tree_tract" &
                                                                          df_psg$posterior_probability ==
                                                                          "probability_80"]))

# Function to format the confidence interval with three digits
format_confidence_interval <- function(conf.int) {
  return(sprintf("[%.3f, %.3f]", conf.int[1], conf.int[2]))
}

# Calculate the ratio of variances
ratio_of_variances_psg <- c(var_test_90_psg$estimate, var_test_80_psg$estimate, log_var_test_90_psg$estimate, log_var_test_80_psg$estimate)

# Create a data frame with test results, including the variance ratio
results_psg <- data.frame(
  "Posterior Probability" = c("90%", "80%", "log(90%)", "log(80%)"),
  "Variance Ratio" = ratio_of_variances_psg,
  "p value" = c(var_test_90_psg$p.value, var_test_80_psg$p.value, log_var_test_90_psg$p.value, log_var_test_80_psg$p.value),
  "Confidence Interval" = sapply(list(var_test_90_psg$conf.int, var_test_80_psg$conf.int, log_var_test_90_psg$conf.int, log_var_test_80_psg$conf.int), format_confidence_interval)
)

# Create a knitr table
kable(results_psg, format = "markdown", digits = 3, align ="c", col.names = c("Posterior Probability", "Variance Ratio", "p", "Confidence Interval"))

```

### Welch Two Sample t-test
``` {r, echo=FALSE, message=FALSE, warning=FALSE}
subset_90_psg = df_psg[df_psg$posterior_probability == "probability_90", ]
results_90_psg <- t.test(psg_count~dist_type, data=subset_90_psg, paired=FALSE, var.eq=F, alternative="two.sided")

subset_80_psg = df_psg[df_psg$posterior_probability == "probability_80", ]
results_80_psg <- t.test(psg_count~dist_type, data=subset_80_psg, paired=FALSE, var.eq=T, alternative="two.sided")

# Create a data frame to store the t-test results
t_test_results_psg <- data.frame(
  posterior_probability = c("90%", "80%"),
  t = c(results_90_psg$statistic, results_80_psg$statistic),
  df = c(results_90_psg$parameter, results_80_psg$parameter),
  p_value = c(results_90_psg$p.value, results_80_psg$p.value),
  conf_interval = c(paste0("[", round(results_90_psg$conf.int[1], 4), ", ", round(results_90_psg$conf.int[2], 4), "]"),
                    paste0("[", round(results_80_psg$conf.int[1], 4), ", ", round(results_80_psg$conf.int[2], 4), "]")
  )
)

# Create a kable table
kable(t_test_results_psg, format = "markdown", align = "c", digits = 3, col.names = c("Posterior Probability", "t", "df", "p", "Confidence Interval"))
```

## dN, dS, dN/dS for Introgressed Regions vs. Genome-Wide Background
``` {r, echo = FALSE}
introgression_tract_dN_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dN_introgression_tract_90.csv"
species_tree_tract_dN_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dN_species_tree_90.csv"

introgression_tract_dS_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dS_introgression_tract_90.csv"
species_tree_tract_dS_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dS_species_tree_90.csv"

introgression_tract_dNdS_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dNdS_introgression_tract_90.csv"
species_tree_tract_dNdS_90 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dNdS_species_tree_90.csv"

introgression_tract_dN_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dN_introgression_tract_80.csv"
species_tree_tract_dN_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dN_species_tree_80.csv"

introgression_tract_dS_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dS_introgression_tract_80.csv"
species_tree_tract_dS_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dS_species_tree_80.csv"

introgression_tract_dNdS_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dNdS_introgression_tract_80.csv"
species_tree_tract_dNdS_80 = "/Users/matt/Documents/Github/dissertation_chapter_2/data/paml/mean_dNdS_species_tree_80.csv"

# Read the CSV file as a one-dimensional vector
dist_dN_introgression_tract_90 <- scan(introgression_tract_dN_90, what = numeric(), sep = ",")
dist_dN_species_tree_90 <- scan(species_tree_tract_dN_90, what = numeric(), sep = ",")
dist_dS_introgression_tract_90 <- scan(introgression_tract_dS_90, what = numeric(), sep = ",")
dist_dS_species_tree_90 <- scan(species_tree_tract_dS_90, what = numeric(), sep = ",")
dist_dNdS_introgression_tract_90 <- scan(introgression_tract_dNdS_90, what = numeric(), sep = ",")
dist_dNdS_species_tree_90 <- scan(species_tree_tract_dNdS_90, what = numeric(), sep = ",")

dist_dN_introgression_tract_80 <- scan(introgression_tract_dN_80, what = numeric(), sep = ",")
dist_dN_species_tree_80 <- scan(species_tree_tract_dN_80, what = numeric(), sep = ",")
dist_dS_introgression_tract_80 <- scan(introgression_tract_dS_80, what = numeric(), sep = ",")
dist_dS_species_tree_80 <- scan(species_tree_tract_dS_80, what = numeric(), sep = ",")
dist_dNdS_introgression_tract_80 <- scan(introgression_tract_dNdS_80, what = numeric(), sep = ",")
dist_dNdS_species_tree_80 <- scan(species_tree_tract_dNdS_80, what = numeric(), sep = ",")
```

``` {r, echo = FALSE}
df_paml <- data.frame(
  dist_type = factor(c(rep("introgression_tract", 1000), rep("species_tree_tract", 1000), rep("introgression_tract", 1000), rep("species_tree_tract", 1000))),
  mean_dN = c(dist_dN_introgression_tract_90, dist_dN_species_tree_90, dist_dN_introgression_tract_80, dist_dN_species_tree_80),
  mean_dS = c(dist_dS_introgression_tract_90, dist_dS_species_tree_90, dist_dS_introgression_tract_80, dist_dS_species_tree_80),
  mean_dNdS = c(dist_dNdS_introgression_tract_90, dist_dNdS_species_tree_90, dist_dNdS_introgression_tract_80, dist_dNdS_species_tree_80),
  posterior_probability = factor(c(rep("probability_90", 2000), rep("probability_80", 2000)))
  )
  
# Convert data to long format
df_paml_long <- df_paml %>% pivot_longer(cols = c(mean_dN, mean_dS, mean_dNdS),
                               names_to = "metric",
                               values_to = "value")

# Reorder the levels of the metric variable
df_paml_long$metric <- factor(df_paml_long$metric, levels = c("mean_dN", "mean_dS", "mean_dNdS"))

# Reorder the levels of the metric variable
df_paml_long$posterior_probability <- factor(df_paml_long$posterior_probability, levels = c("probability_90", "probability_80"))

```

### Welch Two Sample t-test
``` {r, echo=FALSE, message=FALSE, warning=FALSE}
subset_90_paml = df_paml[ df_paml$posterior_probability == "probability_90", ]
results_90_paml <- t.test(mean_dNdS~dist_type, data=subset_90_paml, paired=FALSE, var.eq=F, alternative="two.sided")

subset_80_paml = df_paml[ df_paml$posterior_probability == "probability_80", ]
results_80_paml <- t.test(mean_dNdS~dist_type, data=subset_80_paml, paired=FALSE, var.eq=T, alternative="two.sided")

# Create a data frame to store the t-test results
t_test_results_paml <- data.frame(
  posterior_probability = c("90%", "80%"),
  t = c(results_90_paml$statistic, results_80_paml$statistic),
  df = c(results_90_paml$parameter, results_80_paml$parameter),
  p_value = c(results_90_paml$p.value, results_80_paml$p.value),
  conf_interval = c(paste0("[", round(results_90_paml$conf.int[1], 4), ", ", round(results_90_paml$conf.int[2], 4), "]"),
                    paste0("[", round(results_80_paml$conf.int[1], 4), ", ", round(results_80_paml$conf.int[2], 4), "]")
  )
)

# Create a kable table
kable(t_test_results_paml, format = "markdown", align = "c", digits = 3, col.names = c("Posterior Probability", "t", "df", "p", "Confidence Interval"))
```

### Figure for Manuscript
``` {r, echo = FALSE}
fig <- ggplot(df_paml_long, aes(x = value, fill = dist_type)) +
  geom_density(aes(y=after_stat(density)), alpha=0.8) +
  facet_grid2(
    posterior_probability ~ metric,
    scales = "free", independent = "all",
    labeller = labeller(posterior_probability = c(
      "probability_90" = "90% threshold",
      "probability_80" = "80% threshold"),
      metric = c(
        mean_dN = "Mean dN",
        mean_dS = "Mean dS",
        mean_dNdS = "Mean dNdS"
      )
    ), switch = "x") +
  scale_fill_manual(values = c("introgression_tract" = "#A6CEE3", "species_tree_tract" = "#999999"),
                    labels = c("Introgression Tract", "Genome-Wide Background")) +
  labs(x = "", y = "Probability Density", fill = "") +
  theme_bw() +
  theme(strip.placement = "outside",
        axis.title = element_text(size = 20),
        legend.position = "top",
        panel.grid = element_blank(),
        strip.text.y = element_text(size = 14, angle = 270),
        strip.text.x = element_text(size = 18), # Font size for facet labels
        legend.text = element_text(size = 16),
        axis.text = element_text(size = 8)
  ) +
  guides(fill = guide_legend(title = NULL))
```

```{r, echo=FALSE, message=FALSE, fig.cap = "Figure X. mean H. pulcherrimus - S. fragilis dN, dS, and dN/dS for introgression tracts greater than 10 kb in length and the genome-wide background."}
plot(fig)
```








